### PR TITLE
Add per-DAG delete permissions

### PR DESCRIPTION
**Pull request by blag**
This PR adds per-DAG delete permissions and extends the `sync-perms` subcommand to add delete permissions to the database for all existing DAGs (it does not, however, grant any of the new DAG delete permissions to any roles).

TODO:
* [x] Tests
* [x] ~Documentation~

### CODE REVIEW

**Comment by mik-laj** [comment]
Can you update docs? 
https://airflow.apache.org/docs/apache-airflow/stable/security/access-control.html#dag-level-permissions

File: airflow/www/security.py [hunk 1]
```diff
@@ -325,6 +325,11 @@ def can_edit_dag(self, dag_id, user=None) -> bool:
         dag_resource_name = permissions.resource_name_for_dag(dag_id)
         return self.has_access(permissions.ACTION_CAN_EDIT, dag_resource_name, user=user)
 
+    def can_delete_dag(self, dag_id, user=None) -> bool:
+        """Determines whether a user has DAG delete access."""
+        dag_resource_name = permissions.resource_name_for_dag(dag_id)
+        return self.has_access(permissions.ACTION_CAN_DELETE, dag_resource_name, user=user)
```
Comment (kaxil) [inline airflow/www/security.py:331]: Comment on L352 might need changing i.e might need adding `can_delete`

File: airflow/www/views.py [hunk 2]
```diff
@@ -612,9 +612,12 @@ def add_user_permissions_to_dag(sender, template, context, **extra):
...
         dag.can_trigger = dag.can_edit and can_create_dag_run
-        dag.can_delete = current_app.appbuilder.sm.has_access(
-            permissions.ACTION_CAN_DELETE,
-            permissions.RESOURCE_DAG,
+        dag.can_delete = (
+            current_app.appbuilder.sm.has_access(
+                permissions.ACTION_CAN_DELETE,
+                permissions.RESOURCE_DAG,
+            )
+            or current_app.appbuilder.sm.can_delete_dag(dag.dag_id)
```
Comment (kaxil) [inline airflow/www/views.py]:
```suggestion
        dag.can_delete = current_app.appbuilder.sm.can_delete_dag(dag.dag_id)
```

https://github.com/apache/airflow/blob/1739c61ad13abe9ca3388b7c5faffc6f9854fbc1/airflow/www/security.py#L328-L331

https://github.com/apache/airflow/blob/1739c61ad13abe9ca3388b7c5faffc6f9854fbc1/airflow/www/security.py#L349

https://github.com/apache/airflow/blob/1739c61ad13abe9ca3388b7c5faffc6f9854fbc1/airflow/www/security.py#L365-L368

will already take care of that

**Review by kaxil — COMMENTED** [review]
Approach looks good to me

**Comment by blag** [comment]
@mik-laj Can you be more specific with that you'd like me to document here? As far as I can tell, this PR is really just making delete permissions behave the way they are already implied by the documentation to behave.

File: tests/www/views/test_views_acl.py [hunk 3]
```diff
@@ -190,11 +190,12 @@ def test_permission_exist(acl_app):
     perms_views = acl_app.appbuilder.sm.get_resource_permissions(
         acl_app.appbuilder.sm.get_resource('DAG:example_bash_operator'),
     )
-    assert len(perms_views) == 2
+    assert len(perms_views) == 3
 
     perms = {str(perm) for perm in perms_views}
     assert "can read on DAG:example_bash_operator" in perms
     assert "can edit on DAG:example_bash_operator" in perms
+    assert "can delete on DAG:example_bash_operator" in perms
```
Comment (kaxil) [inline tests/www/views/test_views_acl.py:198]: We probably need one more test as we have in https://github.com/apache/airflow/pull/20346/files#diff-81ef48c7266f1690897901984c78693b82eaef73c99d54c738782fc9541202ceR211-R219

to be sure, that User can:

a) delete a DAG if they just have `DAG.can_delete` 
b) delete a DAG if they just have `DAG:<dag_id>.can_delete` (and not `DAG.can_delete` )

**Review by kaxil — APPROVED** [review]

**Pull request merged**